tools/cp-utility/src/main.rs

```diff
@@ -0,0 +1,338 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+use std::collections::VecDeque;
+use std::env;
+use std::fs;
+use std::io;
+use std::os::unix;
+use std::path::Path;
+use std::path::PathBuf;
+use std::process;
+
+/// A type of copy operation
+#[derive(Debug, PartialEq)]
+enum CopyType {
+    /// equivalent to cp <source> <dest>
+    SingleFile,
+    /// equivalent to cp -a <source> <dest>
+    Archive,
+}
+
+/// Encapsulate a copy operation
+struct CopyOperation {
+    /// The source path
+    source: PathBuf,
+    /// The destination path
+    destination: PathBuf,
+    /// The type of copy being performed
+    copy_type: CopyType,
+}
+
+/// Parse command line arguments and transform into `CopyOperation`
+fn parse_args(args: Vec<String>) -> io::Result<CopyOperation> {
+    if !((args.len() == 4 || args.len() == 5 && args[2].eq("-a")) && args[1].eq("cp")) {
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            "Invalid parameters. Expected cp [-a] <source> <destination>",
+        ));
+    }
+
+    if args.len() == 5 {
+        return Ok(CopyOperation {
+            source: PathBuf::from(&args[3]),
+            destination: PathBuf::from(&args[4]),
+            copy_type: CopyType::Archive,
+        });
+    }
+
+    Ok(CopyOperation {
+        source: PathBuf::from(&args[2]),
+        destination: PathBuf::from(&args[3]),
+        copy_type: CopyType::SingleFile,
+    })
+}
+
+/// Execute the copy operation
+fn do_copy(operation: CopyOperation) -> io::Result<()> {
+    match operation.copy_type {
+        CopyType::Archive => copy_recursive(&operation.source, &operation.destination)?,
+        CopyType::SingleFile => fs::copy(&operation.source, &operation.destination).map(|_| ())?,
+    };
+    Ok(())
+}
+
+/// Execute the recursive type of copy operation
+fn copy_recursive(source: &Path, dest: &Path) -> io::Result<()> {
+    // This will cover the case in which the destination exists
+    let sanitized_dest: PathBuf = if dest.exists() {
+        dest.to_path_buf()
+            .join(source.file_name().ok_or(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "Invalid source
+        file",
```

Comment:
```suggestion
                "Invalid source file",
```

Reply:
I'm using rustfmt. I will try to play around with it to check if we can avoid this.